### PR TITLE
Fabo/export sign message

### PR DESCRIPTION
**Pull request by faboweb**
Exports `createSignMessage` and `createSignedTransaction` for simple use with any message

Cleaned repo, added linting again and fixed testing.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,4 +46,68 @@ const STARGATE_URL = "https://stargate.cosmos.network"
...
+  fee: { amount: [{ denom: ``, amount: `0` }], gas: `21906` },
+  signatures: null,
+  memo: ``
+}
+
+const signMessage = createSignMessage(stdTx, { sequence, accountNumber, chainId });
+const signature = signWithPrivateKey(signMessage, Buffer.from(wallet.privateKey, 'hex'));
+```
+
+### Create a ready to broadcast signed transaction from a message
```
Comment (faboweb) [inline README.md]: @AlexBHarley I added this for you. Would this help you?

**Comment by jnaviask** [comment]
Does this permit signing arbitrary blobs, or just transactions?

**Comment by faboweb** [comment]
> Does this permit signing arbitrary blobs, or just transactions?

`signWithPrivateKey` allows for signing an arbitrary message/blob. But that would need a Backend who handles that. Can you give me a use case so I can give a better evaluation?

File: README.md (hunk 1 above)
Comment (AlexBHarley) [reply README.md]: Yes this is great thank you

File: README.md [hunk 2]
```diff
@@ -46,4 +46,68 @@ const STARGATE_URL = "https://stargate.cosmos.network"
 const api = API(STARGATE_URL)
 
 const validators = await api.validators()
+```
+
+### Create a sign message to sign with on a Ledger or with any other signer
```
Comment (jbibla) [inline README.md]:
```suggestion
### Create a message to be signed by a Ledger or with any other signer app
```

Comment (jbibla) [reply README.md]: does this make sense / is my suggestion true? it was hard to understand precisely what you meant.

File: README.md (hunk 1 above)
Comment (jbibla) [inline README.md]:
```suggestion
### Create and sign a message that is ready for broadcasting
```

Comment (jbibla) [reply README.md]: does a message become a transaction when it's signed? 🤔

File: package.json [hunk 3]
```diff
@@ -5,6 +5,7 @@
   "main": "lib/cosmos.js",
   "scripts": {
     "test": "jest",
+    "lint": "eslint src/* --fix",
```
Comment (jbibla) [inline package.json]: i don't want to add this to your plate right now, but we should consider making a template repo for lunie projects where the package.json scripts, dev dependencies, linting and testing configurations are the same across all our projects.

File: src/messages.js [hunk 4]
```diff
@@ -29,35 +29,35 @@ export function MsgDelegate(
...
   }
 }
 
-export function MsgRedelegate(
+export function MsgRedelegate (
   senderAddress,
   {
-    validator_src_address,
-    validator_dst_address,
+    validatorSrcAddress,
```
Comment (jbibla) [inline src/messages.js]: no abbrv

File: src/messages.js [hunk 5]
```diff
@@ -145,20 +145,20 @@ export function MsgWithdrawDelegationReward(
...
-  "MsgDelegate": MsgDelegate,
-  "MsgUndelegate": MsgUndelegate,
-  "MsgRedelegate": MsgRedelegate,
-  "MsgSubmitProposal": MsgSubmitProposal,
-  "MsgVote": MsgVote,
-  "MsgDeposit": MsgDeposit,
-  "MsgWithdrawDelegationReward": MsgWithdrawDelegationReward
-}
\ No newline at end of file
+  'MsgSend': MsgSend,
```
Comment (jbibla) [inline src/messages.js]: i think our standard is backticks, not single quotes

File: src/signature.js [hunk 6]
```diff
@@ -10,7 +10,7 @@ type StdSignMsg struct {
   Memo          string      `json:"memo"`
 }
 */
-export function createSignMessage(
+export function createSignMessage (
```
Comment (jbibla) [inline src/signature.js]: `signMessage` or `signedMessage`?

File: README.md (hunk 1 above)
Comment (faboweb) [reply README.md]: a transaction is a message wrapped in some meta data ;)

Comment (faboweb) [reply README.md]:
```suggestion
### Create and sign a transaction from a message which then is ready to be broadcast
```

File: package.json (hunk 3 above)
Comment (faboweb) [reply package.json]: Already looked into that but I didn't want to work on this right now :)

File: src/messages.js (hunk 5 above)
Comment (faboweb) [reply src/messages.js]: backticks don't work in object property keys.

File: src/messages.js (hunk 4 above)
Comment (faboweb) [reply src/messages.js]: I can change it

File: src/signature.js (hunk 6 above)
Comment (faboweb) [reply src/signature.js]: `signMessage` so the message you sign

**Comment by jnaviask** [comment]
> `signWithPrivateKey` allows for signing an arbitrary message/blob. But that would need a Backend who handles that. Can you give me a use case so I can give a better evaluation?

We want to verify that a particular client has ownership of a public key, for use in some form of "account management" tool. We don't want to transmit private key/mnemonic information over the wire. This verification can be done through the following algorithm:
1. The client sends us a public key
2. We generate a random token and send it to the client
3. The client uses some offline method (`signWithPrivateKey`?) to sign the token and sends us the signature
4. We verify (through some crypto function?) that the signature matches the pubkey + token, and then mark the client as owner of that public key

Would this work?

**Comment by faboweb** [comment]
> Would this work?

That's how I would do it.

**Comment by jnaviask** [comment]
I just realized you also released this library, so I might not need this feature ported at all https://github.com/luniehq/cosmos-keys

File: src/send.js [hunk 7]
```diff
@@ -1,21 +1,13 @@
...
-  const stdTx = createStdTx({ gas, gasPrices, memo }, messages)
-  const signMessage = createSignMessage(stdTx, { sequence, accountNumber, chainId })
-  let signature, publicKey
-  try {
-    ({ signature, publicKey } = await signer(signMessage))
-  } catch (err) {
-    throw new Error("Signing failed: " + err.message)
-  }
+export default async function send ({ gas, gasPrices = DEFAULT_GAS_PRICE, memo = `` }, messages, signer, cosmosRESTURL, chainId, accountNumber, sequence) {
+  const signedTx = createSignedTransaction({ gas, gasPrices: DEFAULT_GAS_PRICE, memo: `` }, messages, signer, chainId, accountNumber, sequence)
```
Comment (AlexBHarley) [inline src/send.js]: @faboweb createSignedTransaction should be `await`'d. 

`const signedTx = await createSignedTransaction(...`

Have fixed this in my local development branch.

**Pull request merged**